packages/typescript/internal/worker/worker_adapter.ts

```diff
@@ -0,0 +1,162 @@
+ * watcher.
+ */
+let consolidateChangesCallback: ((...args: any[]) => void) | undefined;
+let cachedWatchProgramArgs: string | undefined;
+
+function getWatchProgram(
+  args: string[]
+): ts.WatchOfConfigFile<ts.EmitAndSemanticDiagnosticsBuilderProgram> {
+  const newWatchArgs = args.join(" ");
+
+  // Check to see if the watch program needs to be updated or if we can re-use the old one.
+  if (
+    cachedWatchedProgram &&
+    cachedWatchProgramArgs &&
+    cachedWatchProgramArgs !== newWatchArgs
+  ) {
+    cachedWatchedProgram.close();
+    cachedWatchedProgram = undefined;
+    cachedWatchProgramArgs = undefined;
+  }
+
+  // If we have not yet created a watch
+  if (!cachedWatchedProgram) {
+    const parsedArgs = ts.parseCommandLine(args);
+    const tsconfigPath = args[args.indexOf("--project") + 1];
+
+    cachedWatchProgramArgs = newWatchArgs;
+    cachedWatchedProgram = createWatchProgram(
+      parsedArgs.options,
+      tsconfigPath,
+      (callback) => {
+        consolidateChangesCallback = callback;
+      }
+    );
+  }
+
+  return cachedWatchedProgram;
+}
+
+function emitOnce(args: string[]) {
+  const watchProgram = getWatchProgram(args);
+
+  if (consolidateChangesCallback) {
+    consolidateChangesCallback();
+  }
+
+  return new Promise((res) => {
+    workerRequestTimestamp = Date.now();
+    const result = watchProgram?.getProgram().emit(undefined, undefined, {
+      isCancellationRequested: function(timestamp: number) {
+        return timestamp !== workerRequestTimestamp;
+      }.bind(null, workerRequestTimestamp),
+      throwIfCancellationRequested: function(timestamp: number) {
+        if (timestamp !== workerRequestTimestamp) {
+          throw new ts.OperationCanceledException();
+        }
+      }.bind(null, workerRequestTimestamp),
+    });
+
+    res(result && result.diagnostics.length === 0);
+  });
+}
+
+function main() {
+  const typescriptModuleName = process.argv[process.argv.indexOf('--typescript_module_name') + 1];
+  try {
+    const customTypescriptModule = require(typescriptModuleName);
+    createWatchCompilerHost = customTypescriptModule.createWatchCompilerHost;
+  } catch (e) {
+    worker.log('typescript_module_name must be a requirable module path')
+    throw e
+  }
+  if (process.argv.includes("--persistent_worker")) {
+    worker.log(`Running ${MNEMONIC} as a Bazel worker`);
+    worker.runWorkerLoop(emitOnce);
+  } else {
+    worker.log(`Running ${MNEMONIC} as a standalone process`);
+    worker.log(
+      `Started a new process to perform this action. Your build might be misconfigured, try	
+      --strategy=${MNEMONIC}=worker`
+    );
+
+    let argsFilePath = process.argv.pop()!;
+    if (argsFilePath.startsWith('@')) {
+      argsFilePath = argsFilePath.slice(1)
+    }
+    const args = fs.readFileSync(argsFilePath).toString().split('\n');
+    emitOnce(args).finally(() =>
+      cachedWatchedProgram?.close()
+    );
+  }
+}
+
+main();
```

Comment:
nit: make it testable with `if (require.main === module)` ?

Reply:
Fixed!